File: src/aijack/defense/dp/rdp.cpp
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/functional.h>
#include <cmath>
#include <iostream>
#include <vector>
#include "utils.cpp"

using namespace std;
namespace py = pybind11;

long double eps_gaussian(long double alpha, py::dict params)
{
    if (std::isinf(alpha))
    {
        return min(4 * std::exp(eps_gaussian(2, params) - 1),
                   2 * std::exp(eps_gaussian(2, params)));
    }
    return alpha / (2 * pow(params["sigma"].cast<long double>(), 2));
}

long double eps_laplace(long double alpha, py::dict params)
{
    long double b = params["b"].cast<long double>();
    if (alpha <= 1)
    {
        return 1 / b + std::exp(-1 / b) - 1;
    }
    else if (std::isinf(alpha))
    {
        return 1 / b;
    }
    else
    {
        std::vector<long double> val(2);
        std::vector<int> sgn(2);
        val[0] = std::log(alpha / (2 * alpha - 1)) + (alpha - 1) / b;
        sgn[0] = 1;
        val[1] = std::log((alpha - 1) / (2 * alpha - 1)) + (-1 * alpha) / b;
        sgn[1] = 1;
        return (1 / (alpha - 1)) * logsumexp(val, sgn);
    }
}

long double culc_upperbound_of_rdp_with_Sampled_Gaussian_Mechanism_int(int alpha,
                                                                       py::dict params,
                                                                       long double sampling_rate)
{
    long double sigma = params["sigma"].cast<long double>();
    long double inv_double_sigma_square = 1 / (2 * (sigma * sigma));
    long double log_a = -1 * std::numeric_limits<long double>::infinity();

    for (int i = 0; i <= alpha; i++)
    {
        long double log_coef_i = std::log(binom(alpha, i)) +
                                 i * std::log(sampling_rate) +
                                 (alpha - i) * std::log(1 - sampling_rate);

        long double s = log_coef_i + (i * i - i) * inv_double_sigma_square;
        log_a = _log_add(log_a, s);
    }
    return log_a / (alpha - 1);
}

long double culc_upperbound_of_rdp_with_Sampled_Gaussian_Mechanism_double(long double alpha,
                                                                          py::dict params,
                                                                          long double sampling_rate)
{
    long double sigma = params["sigma"].cast<long double>();
    long double inv_double_sigma_square = 1 / (2 * (sigma * sigma));
    long double log_sampling_rate = std::log(sampling_rate);
    long double log_1m_sampling_rate = std::log(1 - sampling_rate);
    long double log_half = std::log(0.5);

    long double log_a0 = -1 * std::numeric_limits<long double>::infinity();
    long double log_a1 = -1 * std::numeric_limits<long double>::infinity();
    long double i = 0;

    long double z0 = pow(sigma, 2) * std::log(1 / sampling_rate - 1) + 0.5;

    long double coef, log_coef, j, log_t0, log_t1, log_e0, log_e1, log_s0, log_s1;
    while (true)
    {
        coef = binom(alpha, i);
        log_coef = std::log(std::abs(coef));
        j = alpha - i;

        log_t0 = log_coef + i * log_sampling_rate + j * log_1m_sampling_rate;
        log_t1 = log_coef + j * log_sampling_rate + i * log_1m_sampling_rate;

        log_e0 = log_half + _log_erfc((i - z0) / (std::sqrt(2) * sigma));
        log_e1 = log_half + _log_erfc((z0 - j) / (std::sqrt(2) * sigma));

        log_s0 = log_t0 + (i * i - i) * inv_double_sigma_square + log_e0;
        log_s1 = log_t1 + (j * j - j) * inv_double_sigma_square + log_e1;

        if (coef > 0)
        {
            log_a0 = _log_add(log_a0, log_s0);
            log_a1 = _log_add(log_a1, log_s1);
        }
        else
        {
            log_a0 = _log_sub(log_a0, log_s0);
            log_a1 = _log_sub(log_a1, log_s1);
        }

        i += 1;
        if (max(log_s0, log_s1) < -30)
        {
            break;
        }
    }

    return _log_add(log_a0, log_a1) / (alpha - 1);
}

long double culc_upperbound_of_rdp_with_Sampled_Gaussian_Mechanism(long double alpha,
                                                                   py::dict params,
                                                                   long double sampling_rate,
                                                                   const std::function<long double(long double, py::dict)> &_eps)
{
    if (fmod(alpha, 1) == 0.0)
    {
        return culc_upperbound_of_rdp_with_Sampled_Gaussian_Mechanism_int(
            (int)alpha, params, sampling_rate);
    }
    else
    {
        return culc_upperbound_of_rdp_with_Sampled_Gaussian_Mechanism_double(
            alpha, params, sampling_rate);
    }
}

long double culc_tightupperbound_lowerbound_of_rdp_with_theorem6and8_of_zhu_2019(int alpha,
                                                                                 py::dict params,
                                                                                 long double sampling_rate,
                                                                                 const std::function<long double(long double, py::dict)> &_eps)
{
    std::vector<long double> terms(alpha);
    std::vector<int> signs(alpha);

    long double first = ((alpha - 1) * std::log(1 - sampling_rate)) +
                        std::log(alpha * sampling_rate - sampling_rate + 1);
    terms[0] = first;
    signs[0] = 1;

    if (alpha >= 2)
    {
        long double second;
        for (int el = 2; el <= alpha; el++)
        {
            second = (std::log(binom(alpha, el)) +
                      (el * std::log(sampling_rate)) +
                      ((alpha - el) * std::log(1 - sampling_rate)) +
                      ((el - 1) * _eps(el, params)));
            terms[el - 1] = second;
            signs[el - 1] = 1;
        }
    }

    return (1 / ((long double)alpha - 1)) * logsumexp(terms, signs);
}

File: src/aijack/defense/dp/manager/rdp.py
import math

import numpy as np
from scipy import special
from scipy.special import logsumexp

from .utils import _log_add, _log_erfc, _log_sub


def eps_gaussian(alpha, params):
    if alpha == math.inf:
        return min(
            4 * (np.exp(eps_gaussian(2, params) - 1)),
            2 * np.exp(eps_gaussian(2, params)),
        )
    return alpha / (2 * (params["sigma"] ** 2))


def eps_laplace(alpha, params):
    if alpha <= 1:
        return 1 / params["b"] + np.exp(-1 / params["b"]) - 1
    elif alpha == math.inf:
        return 1 / params["b"]
    return (1 / (alpha - 1)) * logsumexp(
        [
            np.log(alpha / (2 * alpha - 1)) + (alpha - 1) / params["b"],
            np.log((alpha - 1) / (2 * alpha - 1)) + (-alpha) / params["b"],
        ],
        b=[1, 1],
    )


def eps_randresp(alpha, params):
    if params["p"] == 1 or params["p"] == 0:
        return math.inf
    if alpha <= 1:
        return (2 * params["p"] - 1) * np.log(params["p"] / (1 - params["p"]))
    elif alpha == math.inf:
        return np.abs(np.log((1.0 * params["p"] / (1 - params["p"]))))
    return (1 / (alpha - 1)) * logsumexp(
        [
            alpha * np.log(params["p"]) + (1 - alpha) * np.log(1 - params["p"]),
            alpha * np.log(1 - params["p"]) + (1 - alpha) * params["p"],
        ],
        b=[1, 1],
    )


def culc_upperbound_of_rdp_with_Sampled_Gaussian_Mechanism(
    alpha, params, sampling_rate, _eps
):
    """Compute log(A_alpha) for any positive finite alpha."""
    if float(alpha).is_integer():
        return culc_upperbound_of_rdp_with_Sampled_Gaussian_Mechanism_int(
            int(alpha), params, sampling_rate
        )
    else:
        return culc_upperbound_of_rdp_with_Sampled_Gaussian_Mechanism_float(
            alpha, params, sampling_rate
        )


def culc_upperbound_of_rdp_with_Sampled_Gaussian_Mechanism_int(
    alpha, params, sampling_rate
):
    """Renyi Differential Privacy of the Sampled Gaussian Mechanism
    3.3 Numerically Stable Computatio
    """
    log_a = -np.inf

    for i in range(alpha + 1):
        log_coef_i = (
            np.log(special.binom(alpha, i))
            + i * np.log(sampling_rate)
            + (alpha - i) * np.log(1 - sampling_rate)
        )

        s = log_coef_i + (i * i - i) / (2 * (params["sigma"] ** 2))
        log_a = _log_add(log_a, s)

    return float(log_a) / (alpha - 1)


def culc_upperbound_of_rdp_with_Sampled_Gaussian_Mechanism_float(
    alpha, params, sampling_rate
):
    """Compute log(A_alpha) for fractional alpha. 0 < q < 1."""
    # The two parts of A_alpha, integrals over (-inf,z0] and [z0, +inf), are
    # initialized to 0 in the log space:
    log_a0, log_a1 = -np.inf, -np.inf
    i = 0

    z0 = params["sigma"] ** 2 * np.log(1 / sampling_rate - 1) + 0.5

    while True:  # do ... until loop
        coef = special.binom(alpha, i)
        log_coef = np.log(abs(coef))
        j = alpha - i

        log_t0 = log_coef + i * np.log(sampling_rate) + j * np.log(1 - sampling_rate)
        log_t1 = log_coef + j * np.log(sampling_rate) + i * np.log(1 - sampling_rate)

        log_e0 = np.log(0.5) + _log_erfc((i - z0) / (math.sqrt(2) * params["sigma"]))
        log_e1 = np.log(0.5) + _log_erfc((z0 - j) / (math.sqrt(2) * params["sigma"]))

        log_s0 = log_t0 + (i * i - i) / (2 * (params["sigma"] ** 2)) + log_e0
        log_s1 = log_t1 + (j * j - j) / (2 * (params["sigma"] ** 2)) + log_e1

        if coef > 0:
            log_a0 = _log_add(log_a0, log_s0)
            log_a1 = _log_add(log_a1, log_s1)
        else:
            log_a0 = _log_sub(log_a0, log_s0)
            log_a1 = _log_sub(log_a1, log_s1)

        i += 1
        if max(log_s0, log_s1) < -30:
            break

    return _log_add(log_a0, log_a1) / (alpha - 1)


def culc_upperbound_of_rdp_with_theorem27_of_wang_2019(
    alpha, params, sampling_rate, _eps
):
    def B(el):
        res = 0
        for i in range(el + 1):
            res += (
                ((-1) ** (i)) * special.binom(el, i) * np.exp((i - 1) * _eps(i, params))
            )
        return abs(res)

    """
    def logAbsB(el):
        ts = []
        ss = []
        for i in range(el + 1):
            ts.append(np.log(special.binom(el, i)) + (i - 1) * _eps(i, params))
            ss.append((-1) ** (i + el % 2))

        return logsumexp(ts, b=ss)
    """

    terms = []
    signs = []

    terms.append(np.log(1))
    signs.append(1)

    second = (
        2 * np.log(sampling_rate)
        + np.log(special.binom(alpha, 2))
        + min(
            np.log(4) + logsumexp([_eps(2, params), np.log(1)], b=[1, -1]),
            _eps(2, params)
            + min(
                np.log(2),
                2 * logsumexp([_eps(math.inf, params), np.log(1)], b=[1, -1]),
            ),
        )
    )
    terms.append(second)
    signs.append(1)

    for j in range(3, alpha + 1):
        terms.append(
            np.log(4)
            + j * np.log(sampling_rate)
            + np.log(special.binom(alpha, j))
            + (1 / 2)
            * (np.log(B(2 * math.floor(j / 2))) + np.log(B(2 * math.ceil(j / 2))))
        )
        signs.append(1)

    return (1 / (alpha - 1)) * logsumexp(terms, b=signs)


def culc_general_upperbound_of_rdp_with_theorem5_of_zhu_2019(
    alpha, params, sampling_rate, _eps
):

    terms = []
    signs = []

    first = ((alpha - 1) * np.log(1 - sampling_rate)) + np.log(
        alpha * sampling_rate - sampling_rate + 1
    )
    terms.append(first)
    signs.append(1)

    if alpha >= 2:
        second = (
            np.log(special.binom(alpha, 2))
            + (2 * np.log(sampling_rate))
            + ((alpha - 2) * np.log(1 - sampling_rate))
            + _eps(2, params)
        )
        terms.append(second)
        signs.append(1)

    if alpha >= 3:
        for el in range(3, alpha + 1):
            third = np.log(3) + (
                np.log(special.binom(alpha, el))
                + (alpha - el) * np.log(1 - sampling_rate)
                + (el * np.log(sampling_rate))
                + ((el - 1) * _eps(el, params))
            )
            terms.append(third)
            signs.append(1)

    return (1 / (alpha - 1)) * logsumexp(terms, b=signs)


def culc_tightupperbound_lowerbound_of_rdp_with_theorem6and8_of_zhu_2019(
    alpha, params, sampling_rate, _eps
):
    terms = []
    signs = []

    first = ((alpha - 1) * np.log(1 - sampling_rate)) + np.log(
        alpha * sampling_rate - sampling_rate + 1
    )
    terms.append(first)
    signs.append(1)

    if alpha >= 2:
        for el in range(2, alpha + 1):
            second = (
                np.log(special.binom(alpha, el))
                + (el * np.log(sampling_rate))
                + ((alpha - el) * np.log(1 - sampling_rate))
                + ((el - 1) * _eps(el, params))
            )
            terms.append(second)
            signs.append(1)

    return (1 / (alpha - 1)) * logsumexp(terms, b=signs)


def culc_tightupperbound_lowerbound_of_rdp_with_theorem6and8_of_zhu_2019_with_tau_estimation(
    alpha, params, sampling_rate, _eps, tau=10
):

    terms = []
    signs = []

    eps_alpha_minus_tau = _eps(alpha - tau, params)

    first = alpha * np.log(1 - sampling_rate) + logsumexp(
        [np.log(1), -eps_alpha_minus_tau], b=[1, -1]
    )
    terms.append(first)
    signs.append(1)

    second = -eps_alpha_minus_tau + alpha * logsumexp(
        [
            np.log(1),
            np.log(sampling_rate),
            np.log(sampling_rate) + eps_alpha_minus_tau,
        ],
        b=[1, -1, 1],
    )
    terms.append(second)
    signs.append(1)

    for el in range(2, tau):
        third = (
            np.log(special.binom(alpha, el))
            + (alpha - el) * np.log(1 - sampling_rate)
            + el * np.log(sampling_rate)
            + logsumexp(
                [(el - 1) * eps_alpha_minus_tau, (el - 1) * _eps(el, params)], b=[1, -1]
            )
        )
        terms.append(third)
        signs.append(-1)

    for el in range(alpha - tau + 1, alpha + 1):
        fourth = (
            np.log(special.binom(alpha, el))
            + (alpha - el) * np.log(1 - sampling_rate)
            + el * np.log(sampling_rate)
            + logsumexp(
                [(el - 1) * _eps(el, params), (el - 1) * eps_alpha_minus_tau],
                b=[1, -1],
            )
        )
        terms.append(fourth)
        signs.append(1)

    return (1 / (alpha - 1)) * logsumexp(terms, b=signs)
